server/apps/microservices/src/processors/video-transcode.processor.ts

```diff
@@ -72,10 +73,45 @@ export class VideoTranscodeProcessor {
   async runVideoEncode(asset: AssetEntity, savedEncodedPath: string): Promise<void> {
     const config = await this.systemConfigService.getConfig();
 
-    if (config.ffmpeg.transcode == TranscodePreset.ALL) {
+    const transcode = await this.needsTranscoding(asset, config.ffmpeg);
+    if (transcode) {
+      //TODO: If video or audio are already the correct format, don't re-encode, copy the stream
       return this.runFFMPEGPipeLine(asset, savedEncodedPath);
     }
+  }
+
+  async needsTranscoding(asset: AssetEntity, ffmpegConfig: SystemConfigFFmpegDto): Promise<boolean> {
+    switch (ffmpegConfig.transcode) {
+      case TranscodePreset.ALL:
+        return true;
+
+      case TranscodePreset.REQUIRED:
+        {
+          const videoStream = await this.getVideoStream(asset);
+          if (videoStream.codec_name !== ffmpegConfig.targetVideoCodec) {
+            return true;
+          }
+        }
+        break;
+
+      case TranscodePreset.OPTIMAL:
+        {
+          const videoStream = await this.getVideoStream(asset);
+          if (videoStream.codec_name !== ffmpegConfig.targetVideoCodec) {
+            return true;
+          }
```

Comment:
Can you deduplicate this code so it only exists once? You might need to move to if/else for that instead of using a switch, I think.

Reply:
Previous version of this method (before my refactoring) didn't have that duplication. But introducing 3 presets (all/required/optimal) worsened the readability of the method. So @jrasm91 suggested to explicitly split code into 3 branches that  resulted in this duplication. Maybe I misunderstood his intention but code after refactoring seems to me better. We now have 3 distinct branches where it is possible to tune transcoding.

Is it fine for now?

Reply:
I had a short attempt at coming up with a tidier approach, but it's a little less straightforward than I thought. All good!